Compiler args are overridden

[this](https://github.com/sergei-lapin/napt/blob/main/plugins/javac/build.gradle#L16-L20) bit of code currently overwrites any existing arguments. would you be open to a PR to make it additive instead?

errr sorry, I was looking at the wrong place. I think it's [here](https://github.com/sergei-lapin/napt/blob/main/plugins/gradle/src/main/java/com/slapin/napt/NaptGradlePlugin.kt#L55-L66), but I'm less sure now. The reason I filed this is because we see this issue now

```
Caused by: java.lang.IllegalAccessError: class com.google.googlejavaformat.java.JavaInput (in unnamed module @0x47a3f73a) cannot access class com.sun.tools.javac.parser.Tokens$TokenKind (in module jdk.compiler) because module jdk.compiler does not export com.sun.tools.javac.parser to unnamed module @0x47a3f73a
        at com.google.googlejavaformat.java.JavaInput.buildToks(JavaInput.java:349)
        at com.google.googlejavaformat.java.JavaInput.buildToks(JavaInput.java:334)
        at com.google.googlejavaformat.java.JavaInput.<init>(JavaInput.java:276)
        at com.google.googlejavaformat.java.Formatter.getFormatReplacements(Formatter.java:260)
        at com.google.googlejavaformat.java.Formatter.formatSource(Formatter.java:247)
        at com.google.googlejavaformat.java.Formatter.formatSource(Formatter.java:213)
        at com.google.googlejavaformat.java.Formatter.formatSource(Formatter.java:200)
        at com.google.googlejavaformat.java.filer.FormattingJavaFileObject$1.close(FormattingJavaFileObject.java:72)
        at com.squareup.javapoet.JavaFile.$closeResource(JavaFile.java:144)
        at com.squareup.javapoet.JavaFile.writeTo(JavaFile.java:173)
        at dagger.android.processor.ContributesAndroidInjectorGenerator.generate(ContributesAndroidInjectorGenerator.java:124)
        at dagger.android.processor.ContributesAndroidInjectorGenerator.process(ContributesAndroidInjectorGenerator.java:84)
        at dagger.android.processor.ContributesAndroidInjectorGenerator.process(ContributesAndroidInjectorGenerator.java:53)
```

--- a/README.md
+++ b/README.md
@@ -11,9 +11,10 @@ An alternative to KAPT that skips stub generation and hence runs up to 50% faste
 
 *JDK 9+ is required to run this particular set of tools*
 
-Once applied, you cant reference generated code in Kotlin code anymore, so you'd have to write Java `bridge` classes in order to reference Java generated code in  Kotlin sources.
+Once applied, you can't reference generated code in Kotlin code anymore, so you'd have to write Java `bridge` classes in
+order to reference Java generated code in Kotlin sources.
 
-For example, assume we have following Kotlin Dagger 2 Component:
+For example, assume we have the following Kotlin Dagger 2 Component:
 
 ``` Kotlin
 @Component
@@ -27,7 +28,8 @@ interface Component {
 }
 ```
 
-then, in order to reference the generated component from Kotlin code we have to write Java `bridge` that would look like this:
+then, in order to reference the generated component from Kotlin code we have to write Java `bridge` that would look like
+this:
 
 ``` Java
 class ComponentBridge {
@@ -45,6 +47,7 @@ That's it, now you can easily reference this `bridge` from your Kotlin code wher
 You could see an example of usage in [sample](https://github.com/sergei-lapin/napt/blob/main/sample/build.gradle)
 
 ## Download
+
 - javac plugin is distributed through Maven Central
 - Gradle plugin is distributed through Gradle Plugin Portal
 
@@ -66,7 +69,7 @@ plugins {
 }
 ```
 
-then you can replace all of your 
+then you can replace all of your
 
 ``` Gradle
 dependencies {
@@ -82,7 +85,7 @@ dependencies {
 }
 ```
 
-That's it. Enjoy speed the speed up of your annotation processing by ~50%.
+That's it. Enjoy speed the speed-up of your annotation processing by ~50%.
 
 #### Ignore NaptTrigger
 
@@ -90,7 +93,8 @@ Add `NaptTrigger.java` to root `.gitignore`
 
 #### Conflitcting NaptTrigger classes
 
-By default Gradle plugin will generate NaptTrigger with module-named package so the FQ names won't clash, but, just in case, the prefix of NaptTrigger package can be specified like that:
+By default, Gradle plugin will generate NaptTrigger with module-named package so the FQ names won't clash, but, just in
+case, the prefix of NaptTrigger package can be specified like that:
 
 ``` Gradle
 napt {
@@ -108,10 +112,27 @@ class NaptTrigger {
 
 #### Disabling NaptTrigger generation
 
-`NaptTrigger.java` generation can be disabled (e.g. when you already have java sources in your module) by setting corresponding property to `false`:
+`NaptTrigger.java` generation can be disabled (e.g. when you already have java sources in your module) by setting
+corresponding property to `false`:
 
 ``` Gradle
 napt {
     generateNaptTrigger.set(false)
+}
+```
+
+#### Supplying custom JVM arguments to Java compiler
+
+In order for napt to work, Java compilation runs in forked process with arguments defined
+by `com.slapin.napt.JvmArgsStrongEncapsulation`. These arguments could be changed by setting corresponding property
+in `napt` extension.
+
+> **Warning**
+> When defining you own `forkJvmArgs` resulting value **must** be a superset
+> of `com.slapin.napt.JvmArgsStrongEncapsulation`
+
+``` Gradle
+napt {
+    forkJvmArgs.set(listOf("your_custom_arg") + JvmArgsStrongEncapsulation)
 }
 ```

--- a/plugins/gradle/src/main/java/com/slapin/napt/NaptGradleExtension.kt
+++ b/plugins/gradle/src/main/java/com/slapin/napt/NaptGradleExtension.kt
@@ -1,6 +1,14 @@
 package com.slapin.napt
 
+import org.gradle.api.provider.ListProperty
 import org.gradle.api.provider.Property
+
+val JvmArgsStrongEncapsulation =
+    listOf(
+        "--add-opens=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED",
+        "--add-opens=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED",
+        "--add-opens=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED",
+    )
 
 abstract class NaptGradleExtension {
 
@@ -12,7 +20,22 @@ abstract class NaptGradleExtension {
 
     /**
      * Defines package prefix for generated NaptTrigger class. Might be useful if project contains
-     * more than 1 equally named module. Defaults to null.
+     * more than 1 equally named module. Default is not specified.
      */
     abstract val naptTriggerPackagePrefix: Property<String>
+
+    /**
+     * Defines additional JVM arguments for the compiler process. Defaults to
+     * [JvmArgsStrongEncapsulation].
+     *
+     * **Warning**: In case of setting this property defined value **must** include arguments
+     * defined in [JvmArgsStrongEncapsulation], e.g.
+     *
+     * ```
+     * napt {
+     *     forkJvmArgs.set(listOf("your_custom_arg") + JvmArgsStrongEncapsulation)
+     * }
+     * ```
+     */
+    abstract val forkJvmArgs: ListProperty<String>
 }

--- a/plugins/gradle/src/main/java/com/slapin/napt/NaptGradlePlugin.kt
+++ b/plugins/gradle/src/main/java/com/slapin/napt/NaptGradlePlugin.kt
@@ -16,11 +16,17 @@ class NaptGradlePlugin : Plugin<Project> {
     override fun apply(target: Project) {
         with(target) {
             val extension = extensions.create("napt", NaptGradleExtension::class.java)
+            extension.defineDefaults()
             insertCompilerPluginDependency()
             bindTriggerCreation(extension)
-            notifyJavaCompilerAboutPlugin()
+            notifyJavaCompilerAboutPlugin(extension)
             bindTriggerCleaning()
         }
+    }
+
+    private fun NaptGradleExtension.defineDefaults() {
+        generateNaptTrigger.convention(true)
+        forkJvmArgs.convention(JvmArgsStrongEncapsulation)
     }
 
     private fun Project.insertCompilerPluginDependency() {
@@ -43,7 +49,7 @@ class NaptGradlePlugin : Plugin<Project> {
                 )
                 task.projectName.set(name)
                 task.packagePrefix.set(extension.naptTriggerPackagePrefix)
-                task.onlyIf { extension.generateNaptTrigger.getOrElse(true) }
+                task.onlyIf { extension.generateNaptTrigger.get() }
                 task.group = "napt"
                 task.description = "Creates NaptTrigger.java in order to trigger NAPT javac plugin"
             }
@@ -52,29 +58,13 @@ class NaptGradlePlugin : Plugin<Project> {
         }
     }
 
-    private fun Project.notifyJavaCompilerAboutPlugin() {
+    private fun Project.notifyJavaCompilerAboutPlugin(extension: NaptGradleExtension) {
         tasks.withType(JavaCompile::class.java).configureEach { javaCompile ->
             javaCompile.options.compilerArgs.add("-Xplugin:Napt")
-            javaCompile.options.fork(
-                mapOf(
-                    "jvmArgs" to
-                        getJdkModuleOpensList(
-                            "jdk.compiler/com.sun.tools.javac.util",
-                            "jdk.compiler/com.sun.tools.javac.api",
-                            "jdk.compiler/com.sun.tools.javac.main",
-                        )
-                )
-            )
+            javaCompile.options.isFork = true
+            requireNotNull(javaCompile.options.forkOptions.jvmArgs)
+                .addAll(extension.forkJvmArgs.get())
         }
-    }
-
-    private fun getJdkModuleOpensList(vararg modulePackage: String): List<String> {
-        val result = mutableListOf<String>()
-        modulePackage.forEach { pkg ->
-            result.add("--add-opens")
-            result.add("$pkg=ALL-UNNAMED")
-        }
-        return result
     }
 
     private fun Project.bindTriggerCleaning() {

--- a/sample/build.gradle.kts
+++ b/sample/build.gradle.kts
@@ -1,12 +1,17 @@
 @file:Suppress("UnstableApiUsage")
 
+import com.slapin.napt.JvmArgsStrongEncapsulation
+
 plugins {
     id("com.android.application")
     id("org.jetbrains.kotlin.android")
     id("com.sergei-lapin.napt")
 }
 
-napt { naptTriggerPackagePrefix.set("com.slapin.napt") }
+napt {
+    naptTriggerPackagePrefix.set("com.slapin.napt")
+    forkJvmArgs.set(listOf("-Dsome.prop=some-value") + JvmArgsStrongEncapsulation)
+}
 
 android {
     compileSdk = 30

